### PR TITLE
[CycleAnalysis] Methods to verify cycles and their nesting.

### PR DESCRIPTION
**Pull request by ssahasra**
The original implementation provided a simple method to check whether the forest of nested cycles is well-formed. This is now augmented with other methods to check well-formedness of all cycles, either invdividually, or as the entire forest. These will be used by future transforms that modify CycleInfo.

### CODE REVIEW

File: llvm/include/llvm/ADT/GenericCycleImpl.h [hunk 1]
```diff
@@ -119,6 +119,101 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
...
+  }
+
+  if (VisitedBBs.size() != getNumBlocks()) {
+    dbgs() << "The following blocks are unreachable in the cycle: ";
+    for (auto *BB : Blocks) {
+      if (!VisitedBBs.count(BB)) {
+        dbgs() << *BB << "\n";
+      }
+    }
+    assert(false && "Unreachable block in cycle");
```
Comment (arsenm) [inline llvm/include/llvm/ADT/GenericCycleImpl.h:186]: llvm_unreachable

File: llvm/include/llvm/ADT/GenericCycleImpl.h [hunk 2]
```diff
@@ -119,6 +119,101 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
...
+           "Cycle contains function entry block!");
+
+    VisitedBBs.insert(BB);
+  }
+
+  if (VisitedBBs.size() != getNumBlocks()) {
+    dbgs() << "The following blocks are unreachable in the cycle: ";
+    for (auto *BB : Blocks) {
+      if (!VisitedBBs.count(BB)) {
+        dbgs() << *BB << "\n";
```
Comment (arsenm) [inline llvm/include/llvm/ADT/GenericCycleImpl.h:183]: Probably should use printAsOperand, and use space separators? Printing the whole block is a lot

File: llvm/include/llvm/ADT/GenericCycleImpl.h (hunk 1 above)
Comment (ssahasra) [reply llvm/include/llvm/ADT/GenericCycleImpl.h:190]: Doesn't really matter here. The whole function is protected by NDEBUG, and meant to be executed only in a debug build. Also, the pattern here is "if (!condition) { print-stuff(); assert(); }" it's not really saying that a particular path is never taken, just that we need to print things before we assert the condition.

File: llvm/include/llvm/ADT/GenericCycleImpl.h (hunk 2 above)
Comment (ssahasra) [reply llvm/include/llvm/ADT/GenericCycleImpl.h]: Done.

File: llvm/include/llvm/ADT/GenericCycleImpl.h [hunk 3]
```diff
@@ -119,6 +120,104 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
...
+  for (BlockT *Entry : entries()) {
+    assert(Entries.insert(Entry).second); // duplicate entry?
+    assert(contains(Entry));
+  }
+
+  // Setup for using a depth-first iterator to visit every block in the cycle.
+  SmallVector<BlockT *, 8> ExitBBs;
+  getExitBlocks(ExitBBs);
+  df_iterator_default_set<BlockT *> VisitSet;
+  VisitSet.insert(ExitBBs.begin(), ExitBBs.end());
```
Comment (arsenm) [inline llvm/include/llvm/ADT/GenericCycleImpl.h:143]: I assume you can do this insert directly in the constructor

File: llvm/include/llvm/ADT/GenericCycleImpl.h (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/ADT/GenericCycleImpl.h:190]: It's always a stylistic choice, and llvm_unreachable is nicer

Comment (ssahasra) [reply llvm/include/llvm/ADT/GenericCycleImpl.h]: Fixed.

File: llvm/include/llvm/ADT/GenericCycleImpl.h (hunk 3 above)
Comment (ssahasra) [reply llvm/include/llvm/ADT/GenericCycleImpl.h:143]: That constructor exists in SmallPtrSet, but turns out that it is not exposed in df_iterator, which is a derived class. So have to go with the default constructor instead.

**Review by arsenm — APPROVED** [review]

**Pull request merged**